client/src/components/history/history-chat.tsx

```diff
@@ -87,13 +87,28 @@ export function HistoryChat(args: ScrollingQuestionsParams): JSX.Element {
 
   const dispatch = useDispatch();
   const [checked, toggleChecked] = useState<boolean>(true);
+  // const [idxQuestion, setIdxQuestion] = useState<number>(-1);
+
+  interface hashMap {
+    [item: number]: boolean;
+  }
+  const chatMap: hashMap = {};
 
   useEffect(() => {
     animateScroll.scrollToBottom({
       containerId: "chat-thread",
     });
+    console.log(chatMap);
   }, [chatData.messages]);
 
+  function updateHashMap(i: number) {
```

Comment:
this is way off the mark. First of all, never write code with meaningless names like `updateHashMap`

In this case, I think what you want is something more like this:

```typescript

// all of this should go in something like a `use-chat-data` hook distinct from the component tsx
const [visibiltityPrefByChatMessageId, setVisibiltityPrefByChatMessageId] = useState<Record<number, boolean>({});

const hideAllPref = useSelector<State, boolean>(s => // whatever select path);

// functions below exported by the hook for use in the component
function setItemVisibilityPref(chatMsgId: int, pref: boolean): void {
  setVisibiltityPrefByChatMessageId({
      ...visibiltityPrefByChatMessageId,
      chatMsgId: pref
  });
}

// return the visibilty pref for a chat item if one is set
// if none set, return the passed-in default
function getItemVisibilityPref{(chatMsgId: int, dft: boolean): boolean {
  return (chatMsgId in visibiltityPrefByChatMessageId)
    ? boolean(visibiltityPrefByChatMessageId[chatMsgId])
    : dft
}

// then in the component itself, you can have a function
// that determines the visibility given an item and following the rules of visibility, e.g.

function isAnswersVisible(chatMsgId: number): boolean {
  const defaultVisForItem = // true if hideAll pref is false OR if this is the last item
  return getItemVisibilityPref(chatMsgId, defaultVisForItem)
}
```